examples/federated-learning/run.py

```diff
@@ -131,74 +80,124 @@ class DataOwner:
-    return iterator.get_next()
+  return grads
 
-  @tfe.local_computation
-  def compute_gradient(self):
-    """Compute gradient given current model parameters and local data."""
-    with tf.name_scope('data_loading'):
-      x, y = self._build_data_pipeline()
+def securely_aggregate(model_grads):
+  with tf.name_scope('secure_aggregation'):
+    aggregated_model_grads = [
+        tfe.add_n(grads) / len(grads)
+        for grads in model_grads
+    ]
 
-    with tf.name_scope('gradient_computation'):
-      grads = self._build_update_step(x, y)
+  return aggregated_model_grads
 
-    return grads
+@tfe.local_computation
+def validation_step(model_owner):
+  """Runs a validation step!"""
+  x, y = next(model_owner.dataset)
 
+  with tf.name_scope('validate'):
+    predictions = model_owner.model(x)
+    loss = tf.reduce_mean(model_owner.loss(y, predictions, from_logits=True))
+
+    y_hat = tf.argmax(input=predictions, axis=1)
+
+  tf.print("loss", loss)
+  tf.print("expect", y, summarize=50)
+  tf.print("result", y_hat, summarize=50)
+  return loss, y, y_hat
+
+@tfe.local_computation
+def train_step(data_owner):
+  """Runs a single training step!"""
+  x, y = next(data_owner.dataset)
+
+  with tf.name_scope('gradient_computation'):
+    with tf.GradientTape() as tape:
+      predictions = model(x)
+      loss = tf.reduce_mean(data_owner.loss(y, predictions, from_logits=True))
+
+    grads = tape.gradient(loss, model.trainable_variables)
+
+  return grads
+
+
+@tf.function
+def train_step_master(model_owner, data_owners):
+  """Runs a single training step on each data owner!"""
+  grads = []
+
+  for data_owner in data_owners:
+    grads.append(train_step(data_owner, player_name=data_owner.player_name))
+
+  agg_grads = securely_aggregate(zip(*grads))
+
+  update_model(model_owner, *agg_grads, player_name=model_owner.player_name)
+  validation_step(model_owner, player_name=model_owner.player_name)
 
 if __name__ == "__main__":
 
   logging.basicConfig(level=logging.DEBUG)
 
-  model_owner = ModelOwner("model-owner")
+  model = tf.keras.Sequential((
+      tf.keras.layers.Dense(512, input_shape=[None, 28 * 28],
+                            activation='sigmoid'),
+      tf.keras.layers.Dense(10),
+  ))
+
+  model.build()
+
+  loss = tf.keras.losses.sparse_categorical_crossentropy
+
+  model_owner = ModelOwner("model-owner", model,
+                           tf.keras.optimizers.Adam(), loss)
   data_owners = [
-      DataOwner("data-owner-0", "./data/train.tfrecord",
-                model_owner.build_update_step),
-      DataOwner("data-owner-1", "./data/train.tfrecord",
-                model_owner.build_update_step),
-      DataOwner("data-owner-2", "./data/train.tfrecord",
-                model_owner.build_update_step),
+      DataOwner("data-owner-0", "./data/train.tfrecord", loss),
+      DataOwner("data-owner-1", "./data/train.tfrecord", loss),
+      DataOwner("data-owner-2", "./data/train.tfrecord", loss),
```

Comment:
To validate that the model is learning currently, I would suggest that the `data-owner` 0, 1 and 2 are the 'model-owner' are using different dataset (currently always the same `train.tfrecord`). These files could be generated by `download.py`. When generating this dataset, we just need to make sure the MNIST data has been shuffled, in order to avoid that the data-owner-0 got all the 0 and 1 digits, data-owner-1 all the the 1 and 2. thanks